### PR TITLE
Bug 4809 Docker Run Script Fixes and Improvements

### LINKED ISSUE #4809: Docker run script fails on systems without coreutils (realpath)
#### Expected Behavior
Relative paths should normalize at the prompts.

#### Actual Behavior
Crashes because realpath isn't available.

#### Potential Solution
Polyfill.

#### Reproducing the Problem
Run on macOS or old Debian.

#### System Information
macOS Big Sur

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by C-SELLERS**
#### Description

Changes to our Docker run scripts (run_docker.sh/bat & run_docker_notebook.sh/bat):
- Remove dependency for coreutils on Linux/Mac
- Bash scripts handle already running containers
- Mirror .sh changes to .bat scripts to maintain uniform functionality
- Store results in ./Results dir by default
- Update readmes with any functional changes

Special thanks to @kazazes for starting these changes

#### Related Issue

Closes #4809 
Closes #4810 

#### Motivation and Context

Coreutils problem mentioned in #4809 
Some quality of life improvements as well for easy use in linux environment

#### Requires Documentation Change

Readmes have been updated

#### How Has This Been Tested?

Ran a three variations of both scripts on Ubuntu 20.04 and Windows 10

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Research/readme.md [hunk 1]
```diff
@@ -21,11 +21,11 @@ From a terminal launch the run_docker_notebook.bat/.sh script; there are a few o
 
  2. Using the **docker.cfg** to store args for repeated use; any blank entries will resort to default values! ex: `./run_docker_notebook.bat docker.cfg`
   
-        image=
-        data_dir=
-        notebook_dir=
+         IMAGE=quantconnect/lean:latest
```
Comment (Martin-Molinero) [inline Research/readme.md:24]: same here `lean` -> `research`?

File: Research/run_docker_notebook.sh [hunk 2]
```diff
@@ -33,40 +51,56 @@ elif [ ! -z "$*" ]; then
...
+    read -p "Enter absolute path to store notebooks [default: $DEFAULT_NOTEBOOK_DIR]: " NOTEBOOK_DIR
 fi
 
 #Have to reset IFS for cfg files to work properly
 IFS=" "
 
-if [ -z "$image" ]; then
-    image=$default_image
-fi
+# fall back to defaults on empty input without
```
Comment (Martin-Molinero) [inline Research/run_docker_notebook.sh:62]: typo `without` ?

File: run_docker.bat [hunk 3]
```diff
@@ -13,19 +13,19 @@ REM WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
...
-set default_data_dir=%current_dir%Data\
-set default_results_dir=%current_dir%
-set default_config_file=%current_dir%Launcher\config.json
-set default_python_dir=%current_dir%Algorithm.Python\
-set csharp_dll=%current_dir%Launcher\bin\Debug\QuantConnect.Algorithm.CSharp.dll
-set csharp_pdb=%current_dir%Launcher\bin\Debug\QuantConnect.Algorithm.CSharp.pdb
+set CURRENT_DIR=%~dp0
+set DEFAULT_IMAGE=quantconnect/lean:latest
+set DEFAULT_DATA_DIR=%CURRENT_DIR%Data\
+set DEFAULT_RESULTS_DIR=%CURRENT_DIR%\Results
```
Comment (Martin-Molinero) [inline run_docker.bat:19]: `\Results` think `\` isn't required? I'm get `Enter absolute path to store results [default: ...\Lean\\Results]`

File: run_docker.bat [hunk 4]
```diff
@@ -36,96 +36,96 @@ if not "%*"=="" (
...
-    set /p image="Enter docker image [default: %default_image%]: "
-    set /p config_file="Enter absolute path to Lean config file [default: %default_config_file%]: "
-    set /p data_dir="Enter absolute path to Data folder [default: %default_data_dir%]: "
-    set /p results_dir="Enter absolute path to store results [default: %default_results_dir%]: "
-    set /p debugging="Would you like to debug C#? (Requires mono debugger attachment) [default: N]: "	
+    set /p IMAGE="Enter docker image [default: %DEFAULT_IMAGE%]: "
+    set /p CONFIG_FILE="Enter absolute path to Lean config file [default: %DEFAULT_CONFIG_FILE%]: "
+    set /p DATA_DIR="Enter absolute path to Data folder [default: %DEFAULT_DATA_DIR%]: "
+    set /p RESULTS_DIR="Enter absolute path to store results [default: %DEFAULT_RESULTS_DIR%]: "
+    set /p DEBUGGING="Would you like to debug C#? (Requires mono debugger attachment) [default: N]: "
```
Comment (Martin-Molinero) [inline run_docker.bat:43]: `run_docker.sh` allows specifying `Python directory` should we add it here too?

File: run_docker.sh [hunk 5]
```diff
@@ -13,108 +12,143 @@
...
-
-#First part of the docker command that is static, then we build the rest
-command="docker run --rm --mount type=bind,source=$config_file,target=/Lean/Launcher/config.json,readonly \
-    --mount type=bind,source=$data_dir,target=/Data,readonly \
-    --mount type=bind,source=$results_dir,target=/Results \
-    --name LeanEngine \
+#First part of the docker COMMAND that is static, then we build the rest
+COMMAND="docker run --rm -v $CONFIG_PATH:/Lean/Launcher/config.json:ro \
+    -v $DATA_DIR:/Data:ro \
+    -v $RESULTS_DIR:/Results \
```
Comment (Martin-Molinero) [inline run_docker.sh:102]: I think we should normalize the behavior across all docker scripts, whether we use `-v` or `--mount`

File: run_docker.sh [hunk 6]
```diff
@@ -13,108 +12,143 @@
...
+SUDO=""
+
+# verify if user has docker permissions
+if ! touch /var/run/docker.sock &>/dev/null; then
+  sudo -v
+  SUDO="sudo"
+  COMMAND="$SUDO $COMMAND"
+fi
+
+if [ "$($SUDO docker container inspect -f '{{.State.Running}}' $CONTAINER_NAME)" == "true" ]; then
```
Comment (Martin-Molinero) [inline run_docker.sh:143]: I'm getting:
```
root@ns569941:/home/martin/Lean# ./run_docker.sh
Docker image [default: quantconnect/lean:latest]:
Path to Lean config.json [default: ./Launcher/config.json]:
Path to Data directory [default: ./Data]:
Path to Results directory [default: ./Results]:
Path to Python directory [default: ./Algorithm.Python/]:
Would you like to debug C#? (Requires mono debugger attachment) [default: N]:
Results directory /home/martin/Lean/Results does not exist; creating it now
./run_docker.sh: line 145: syntax error near unexpected token `)'
./run_docker.sh: line 145: `    ($SUDO docker stop $CONTAINER_NAME))'
```

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice improvements, leaving some requests

File: run_docker.sh (hunk 6 above)
Comment (C-SELLERS) [reply run_docker.sh:143]: 🤔

**Review by Martin-Molinero — APPROVED** [review]
Changes look good 👍 
Suggest just before the `docker run` we first do `docker pull %IMAGE%` in the different scripts, found a weird behavior that if executing the `run_docker_notebook.bat` using a console (doesn't happen if executed directly) and the latest image is not present it won't pull it and just end immediately, took me while to understand why it was happening 😅

**Comment by C-SELLERS** [comment]
Gotcha! Just added a option to update the image, by default it is set to yes! Just didn't want to force updates if someone is running it offline with local data for backtesting.

**Comment by C-SELLERS** [comment]
Also, thats very weird ? I haven't had that happen when running from command line 🤔 Even with an outdated image. But either way I think the auto update of the image is a good feature to tack on!

**Pull request merged**